File: js/state-monad.js
var stateMonad = new function() {
  var wrap = function(drink) {
    return function(sugars) {
      return {drink: drink,
              sugars: sugars};
    }
  };

  var bind = function(mv, f) {
    return function(sugars) {
      var value     = mv(sugars);
      var drink     = value.drink,
          newSugars = value.sugars;
      return f(drink)(newSugars);
    }
  };
  return {
    wrap: wrap,
    bind: bind
  }
}();

exports.stateMonad = stateMonad;

File: clojure/src/monads_blog/state_monad.clj
(ns monads-blog.state-monad)

(defn wrap[drink]
  (fn[sugars]
    {'drink drink 'sugars sugars}))

(defn bind[mv f]
  (fn[sugars]
    (let [realized  (mv sugars)
          drink     (realized 'drink)
          newSugars (realized 'sugars)]
      ((f drink) newSugars))))

